src/gsIO/gsBenchmark.h

```diff
@@ -0,0 +1,548 @@
+          result = test();
+        }
+
+        stopwatch.stop();
+        runtime = stopwatch.elapsed()/(real_t)runs;
+
+        switch(metric & ~gismo::metric::speedup & ~gismo::metric::ratio) {
+        case gismo::metric::bandwidth_kb_sec: case gismo::metric::perf_kflop_sec:
+          value = 1e-03 * result / runtime;
+          break;
+        case gismo::metric::bandwidth_mb_sec: case gismo::metric::perf_mflop_sec:
+          value = 1e-06 * result / runtime;
+          break;
+        case gismo::metric::bandwidth_gb_sec: case gismo::metric::perf_gflop_sec:
+          value = 1e-09 * result / runtime;
+          break;
+        case gismo::metric::bandwidth_tb_sec: case gismo::metric::perf_tflop_sec:
+          value = 1e-12 * result / runtime;
+          break;
+        case gismo::metric::runtime_sec:
+          value = runtime;
+          break;
+        default:
+          GISMO_ERROR("Unsupported metric");
+        }
+
+        gsBenchmarkResult result;
+        result.threads = static_cast<int>(it); // number of OpenMP threads
+        result.runtime = runtime;              // averaged elapsed time in seconds
+        result.value   = value;                // averaged benchmark value
+        result.metric  = metric;               // benchmark metric
+        results.push_back( give(result) );
+      }
+    } catch(...) {}
+
+    // Convert to relative values (speedup relative to first entry)
+    if (metric & gismo::metric::speedup) {
+      runtime = results.front().runtime;
+      value   = results.front().value;
+
+      for (auto& it : results) {
+        it.runtime = runtime / it.runtime;
+        it.value   = value   / it.value;
+      }
+    }
+
+    return results;
+  }
+
+private:
+  std::vector<gsBenchmarkSet> benchmarks;
+};
+
+/// Print (as string) operator
+inline std::ostream &operator<<(std::ostream &os, const gsBenchmark& obj)
+{ return obj.print(os); }
+
+namespace internal
+{
+/// @brief Get a gsBenchmark from XML data
+template<>
+class gsXml< gsBenchmark >
+{
+private:
+  gsXml() { }
+  typedef gsBenchmark Object;
+public:
+  GSXML_COMMON_FUNCTIONS(Object);
+  static std::string tag ()  { return "Benchmark"; }
+  static std::string type () { return "Benchmark"; }
+
+  GSXML_GET_POINTER(Object);
+
+  static void get_into (gsXmlNode * node, Object & obj)
+  {
+    gsXmlNode * child;
+
+    child = node->first_node(gsXml< gsBenchmarkSet >::tag().c_str());
+    for (; child; child = child->next_sibling() ) {
+      gsBenchmarkSet benchmark;
+      gsXml< gsBenchmarkSet >::get_into(child, benchmark);
+      obj.get().push_back( give(benchmark) );
+    }
+  }
+
+  static gsXmlNode * put (const Object & obj, gsXmlTree & data )
+  {
+    gsXmlNode * results = makeNode("Benchmark", data);
+
+    for (const auto& it : obj.get()) {
+      results->append_node( gsXml< gsBenchmarkSet >::put(it, data) );
+    }
+
+    return results;
+  }
+};
+} // namespace internal
+
+namespace benchmark {
```

Comment:
We could avoid a new namespace, if we add this to gismo::util namespace
and rename ratio to eg.  benchmartRatio.
in that case we call
util::benchmarkRatio

Reply:
I can do that. Would you then move the `ratio()` functions into `gsUtils.h`? I'd prefer to keep all benchmark-related stuff in `gsBenchmark.{h,cpp}`.

Reply:
they can stay in the file, but I only propose to have them in "util" namespace